wgpu-hal/src/dx12/shader_compilation.rs

```diff
@@ -0,0 +1,260 @@
+// fallback to FXC if the Dxc libraries (dxil.dll and dxcompiler.dll) are not found, or if Fxc is chosen at'
+// `Instance` creation time.
+
+pub(super) fn compile_fxc(
+    device: &super::Device,
+    source: &String,
+    source_name: &str,
+    raw_ep: &std::ffi::CString,
+    stage_bit: wgt::ShaderStages,
+    full_stage: String,
+) -> (
+    Result<super::CompiledShader, crate::PipelineError>,
+    log::Level,
+) {
+    let mut shader_data = native::Blob::null();
+    let mut compile_flags = d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS;
+    if device
+        .private_caps
+        .instance_flags
+        .contains(crate::InstanceFlags::DEBUG)
+    {
+        compile_flags |= d3dcompiler::D3DCOMPILE_DEBUG | d3dcompiler::D3DCOMPILE_SKIP_OPTIMIZATION;
+    }
+    let mut error = native::Blob::null();
+    let hr = unsafe {
+        profiling::scope!("d3dcompiler::D3DCompile");
+        d3dcompiler::D3DCompile(
+            source.as_ptr().cast(),
+            source.len(),
+            source_name.as_ptr().cast(),
+            ptr::null(),
+            ptr::null_mut(),
+            raw_ep.as_ptr(),
+            full_stage.as_ptr().cast(),
+            compile_flags,
+            0,
+            shader_data.mut_void().cast(),
+            error.mut_void().cast(),
+        )
+    };
+
+    match hr.into_result() {
+        Ok(()) => (
+            Ok(super::CompiledShader::Fxc(shader_data)),
+            log::Level::Info,
+        ),
+        Err(e) => {
+            let mut full_msg = format!("FXC D3DCompile error ({})", e);
+            if !error.is_null() {
+                use std::fmt::Write as _;
+                let message = unsafe {
+                    std::slice::from_raw_parts(
+                        error.GetBufferPointer() as *const u8,
+                        error.GetBufferSize(),
+                    )
+                };
+                let _ = write!(full_msg, ": {}", String::from_utf8_lossy(message));
+                unsafe {
+                    error.destroy();
+                }
+            }
+            (
+                Err(crate::PipelineError::Linkage(stage_bit, full_msg)),
+                log::Level::Warn,
+            )
+        }
+    }
+}
+
+// The Dxc implementation is behind a feature flag so that users who don't want to use dxc can disable the feature.
+#[cfg(feature = "dxc_shader_compiler")]
+mod shader {
+    use std::path::PathBuf;
+
+    pub(crate) struct DxcContainer {
+        pub compiler: hassle_rs::DxcCompiler,
+        pub library: hassle_rs::DxcLibrary,
+        // Has to be held onto for the lifetime of the device otherwise shaders will fail to compile
+        _dxc: hassle_rs::Dxc,
+    }
+
+    pub(crate) fn get_dxc_container(
+        dxc_path: Option<PathBuf>,
+        dxil_path: Option<PathBuf>,
+    ) -> Result<Option<DxcContainer>, crate::DeviceError> {
+        // Make sure that dxil.dll exists.
+        match hassle_rs::Dxil::new(dxil_path) {
```

Comment:
`dxil.dll` never gets used directly, it just gets sneakily called by `dxcompiler.dll` to validate the shaders during compilation so I didn't bother storing it.